src/start/AuthScreen.js

```diff
@@ -1,24 +1,109 @@
+import { Linking, Platform, ScrollView, Text, View } from 'react-native';
 import { connect } from 'react-redux';
 
+import SafariView from 'react-native-safari-view';
+import parseURL from 'url-parse';
+
 import boundActions from '../boundActions';
 import styles from '../common/styles';
-import { Input, Screen, ZulipButton } from '../common';
+import { Input, Screen, ZulipButton, GoogleButton } from '../common';
+import { generateOtp, extractApiKey } from '../utils/encoding';
 import { getAuth } from '../account/accountSelectors';
-import { getCurrentRoute } from '../nav/routingSelectors';
 
 import PasswordAuthView from './PasswordAuthView';
 
-class AuthScreen extends React.Component {
+class AuthScreen extends React.PureComponent {
 
   props: {
     authBackends: string[],
   };
 
+  componentDidMount = () => {
+    // Add listeners for OAuth flow
+    Linking.addEventListener('url', this.endOAuthFlow);
+    this.safariViewDismissEvent = SafariView.addEventListener('onDismiss', () => {
+      this.otp = undefined;
+    });
+  }
+
+  componentWillUnmount = () => {
+    // Remove listeners for OAuth flow
+    Linking.removeEventListener('url', this.endOAuthFlow);
+    SafariView.removeEventListener('onDismiss', this.safariViewDismissEvent);
+  }
+
   handleTypeSelect = (authType: string) => {
-    this.props.setAuthType(authType);
-    this.props.pushRoute(authType);
+    const { realm } = this.props;
+
+    if (authType === 'google') {
+      // Google OAuth flow
+      this.beginOAuthFlow(`${realm}/accounts/login/google/`);
+    } else if (authType === 'github') {
+      // Github OAuth flow
+      this.beginOAuthFlow(`${realm}/accounts/login/social/github/`);
+    } else {
+      // Password auth flow
+      this.props.setAuthType(authType);
+      this.props.pushRoute(authType);
+    }
+  }
+
+  beginOAuthFlow = async (url) => {
+    // Generate a one time pad (OTP) to send up with the request
+    // The server XORs the API key with this OTP in its response to protect
+    // against malicious apps registering an identical URI scheme and trying
+    // to intercept credentials
+    this.otp = await generateOtp();
+    SafariView.show({ url: `${url}?mobile_flow_otp=${this.otp}` });
+  };
+
+  endOAuthFlow = (event) => {
+    const { realm, loginSuccess } = this.props;
+
+    SafariView.dismiss();
+
+    const url = parseURL(event.url, true);
+
+    // The OAuth callback should have the following format:
+    // zulip://login?realm={}&email={}&otp_encrypted_api_key={}
+    if (url.host === 'login') {
+      const query = url.query;
+
+      // Check for errors
+      if (query.realm !== realm) {
+        // eslint-disable-next-line no-console
+        console.error('Zulip realm does not match request.');
+        return;
+      }
+
+      if (!this.otp) {
+        // eslint-disable-next-line no-console
+        console.error('No one time pad stored. This auth redirect may not have been requested.');
```

Comment:
Oh ok. Perhaps we should enable logging `console.error`? It would be nice if we had a way to log errors that don't crash the app. Either way I think it's nice to have these in for debugging purposes